### ISSUE TITLE
Search engine friendly dataset output

### ISSUE THREAD

**Issue by BigBlueHat** [codeforgreenville/upstate_tech_cal_service#12]
This may be a bit of a lift to make happen, but it would turn the [/api/gtc](https://greenville-cal-service.herokuapp.com/api/gtc) output into something that could be fed to a search engine or graph-based too (see [LevelGraph](http://levelgraph.io) and it's [Playground](https://wileylabs.github.io/levelgraph-playground/)) with meaningful results (rich snippet display, filtering down to events happening at a single place, etc).

Here's an example of the current output (which search engines can't make much out of since it uses a "colloquial" vocabulary--only known to the developers here):
```json
[
  {
    "created_at": "2018-08-09T22:51:43Z", 
    "data_as_of": "2019-01-27T05:43:14Z", 
    "description": "Come and Design, Write Copy, Hack on the Code for Greenville Projects. We won't do much updating during this meeting, it's really around getting down to work. If you are attending for the first time, you may want to start with our monthly Brigade Meeting on 3rd Thursdays. If you want to join in and share your skills this is the meetup for you! Working on:<br/>* Android Tolley Tracker Client App<br/>* iOS Trolley Tracker Agent App*<br/>* Trolley Tracker API<br/>* Open Data API<br/>* Trolley Tracker Design and Marketing *After 6pm the Bank of America elevators are not available to the public so it is best to enter from the 3rd floor of the Richardson Street public parking garage*", 
    "event_name": "Code For Greenville Work Night", 
    "group_name": "Code for Greenville", 
    "nid": "7", 
    "rsvp_count": 0, 
    "tags": "1", 
    "time": "2019-03-05T23:00:00Z", 
    "url": "https://www.meetup.com/Code-for-Greenville/events/mcrjcqyzfbhb/", 
    "uuid": "47500439-5c63-45ba-9600-be41109bc291", 
    "venue": {
      "address": "101 N. Main St 3rd Floor", 
      "city": "Greenville", 
      "country": "us", 
      "lat": 34.851925, 
      "lon": -82.399948, 
      "name": "OpenWorks", 
      "state": "SC", 
      "zip": null
    }
  }
]
```

Here's a revision that's search engine ready and can be used to create a graph for finding cool stuff (yes it's more verbose...because it does more things with the same data 😄):
```json
{
"@context": [
  "http://schema.org/",
  {"nid": null, "rsvp_count": null, "tags": null, "uuid": null}
],
"@type": "DataFeed",
"dataFeedElement": [
  {
    "@type": "DataFeedItem",
    "dateCreated": "2018-08-09T22:51:43Z", 
    "dateModified": "2019-01-27T05:43:14Z",
    "item": {
      "@type": "Event",
      "description": "Come and Design, Write Copy, Hack on the Code for Greenville Projects. We won't do much updating during this meeting, it's really around getting down to work. If you are attending for the first time, you may want to start with our monthly Brigade Meeting on 3rd Thursdays. If you want to join in and share your skills this is the meetup for you! Working on:<br/>* Android Tolley Tracker Client App<br/>* iOS Trolley Tracker Agent App*<br/>* Trolley Tracker API<br/>* Open Data API<br/>* Trolley Tracker Design and Marketing *After 6pm the Bank of America elevators are not available to the public so it is best to enter from the 3rd floor of the Richardson Street public parking garage*", 
      "name": "Code For Greenville Work Night", 
      "organizer": {
        "@type": "Organization",
        "name": "Code for Greenville"
      }, 
      "nid": "7",
      "rsvp_count": 0, 
      "tags": "1", 
      "startDate": "2019-03-05T23:00:00Z", 
      "url": "https://www.meetup.com/Code-for-Greenville/events/mcrjcqyzfbhb/", 
      "identifier": "urn:uuid:47500439-5c63-45ba-9600-be41109bc291", 
      "location": {
        "@type": "Place",
        "name": "OpenWorks",
        "address": {
          "streetAddress": "101 N. Main St 3rd Floor", 
          "addressLocality": "Greenville",
          "addressRegion": "SC",
          "addressCountry": "USA",
          "postalCode": null
        },
        "geo": {
          "@type": "GeoCoordinates",
          "latitude": 34.851925, 
          "longitude": -82.399948
        }
      }
    }
  }
]}
```

In both examples the array would have more objects in it based on the available data of course.

Also, this object in the `@context` object keeps these terms out of the graph data (as search engines and graphs won't know what to do with them anyhow), so you could also just not output those:
```json
  {"nid": null, "rsvp_count": null, "tags": null, "uuid": null}
```

Hope that mapping solves the hard part anyhow! 😃 

Cheers,
🎩

**Comment by allella** [comment]
Thanks. Marked as an enhancement and perhaps we can get this going on a future push.

**Comment by ramonaspence** [comment]
Oh I'd love to do this. And I think if I'm updating the tests and test fixtures, it may be good to do this first.

I'm curious what this would look like with multiple events. Would it simply be multiple dictionaries inside the list/array named `dataFeedElement`?

**Comment by allella** [comment]
It looks like the https://schema.org/DataFeed has an example in the JSON-LD section at the bottom of that page.

That example seems to show how to do multiple [DataFeedItems](https://schema.org/DataFeedItem) inside a DataFeed->DataFeedElements object.

@BigBlueHat we have a few applications already consuming our existing JSON API.  Is the expectation that this is another "view" of the data and not replacing our existing API?  If that's the case, would the search engines, or applications that want the JSON-LD / schema.org context, need to send a specific HTTP header to the API server to indicate which context it wants back?

**Comment by BigBlueHat** [comment]
@allella I'd suggest replacing the current array based one--as arrays as top-level objects don't give you any way to describe what's in the array.

However, given that'd break any existing integrations, the best approach would be to setup the server so it will respond to requests for `application/ld+json` and have that return the `dataFeedElement` above.

**Comment by allella** [comment]
Thanks.  @ramonaspence I'm not sure where in Flask to interrogate the headers, but it seems if the request includes an
`Accept: application/ld+json` HTTP header that we'd return the schema.org / dataFeedElement. Otherwise, we'd continue to return our existing JSON.  Does that make sense?

**Comment by ramonaspence** [comment]
@allella That does make sense. I'm looking into how to implement it now. Flask does have a request object so I can check the request headers and implement the conditional logic from there.

This will be my project for the evenings this week, as I go back to work tomorrow.

**Comment by BigBlueHat** [comment]
You might give [Flask-Arrest](https://pythonhosted.org/Flask-arrest/index.html) or some other content negotiation library a look. They'll help simplify the process of parsing all the possible variations of the Accept header (which can be multi-valued with quality priorities).

A one to one match is a fine start, though.  :)

**Comment by ramonaspence** [comment]
@allella There's one small enhancement that I think would be good to do while we're implementing these two different responses for JSON and LD+Json. 

The view function `get_dates()` does far more than getting date/time parameters. It's the primary and only view function. So I'm wondering, should I go ahead and make the view functions in this app more modular? It would make it more readable and easier for others to work on in the future. 

I think if there were an index view that simply provides JSON format (I think the "check" for LD+json request will go in here) with the default start_date and end_date,  and a separate view to handle additional date parameters.

It may not be the right way to do this for reasons that I'm not aware of but I wanted to pitch the idea.

**Comment by allella** [comment]
This was thrown together by the initial dev and hacked on by a few others, so feel free to clean it up.  As long as it returns the same output you can do what you think is good.

We don't have any tests to verify things won't break, but off the top of my head the main use cases are returning all future events, returning events based on the date range provided, and returning based on tags. I believe you could also do date and tag filters together, but not sure if that works as they may be mutually exclusive.

The README lays out those cases https://github.com/codeforgreenville/upstate_tech_cal_service#interacting-with-a-running-application

**Comment by ramonaspence** [comment]
@BigBlueHat @allella 

EDIT: Sorry for the false alarm. I got help clearing up something in the flask-arrest docs that I didn't understand initially. So now I believe I'll be able to have a PR to review with a couple of unit tests thrown in over the weekend.


...
I believe I'm getting close to completing this. I've been sick for the past week  :(

I'm missing some understanding of content negotiation in general, so I wanted to reach out and make sure I'm understanding what to do with flask-arrest.

As of right now, I have a function that brings in the local data from `all_meetings.json` and reformats it using the schema provided by OP. So, I'm down to checking request headers, and rendering the appropriate response.

With Flask-arrest, it seems like I could check the request headers and then use `serialize_response()`  to serialize for JSON or JSON+LD. Does that sound close for what we're trying to accomplish?
(https://pythonhosted.org/Flask-arrest/content-negotiation.html#flask_arrest.helpers.serialize_response)

**Comment by allella** [comment]
@ramonaspence I believe we'd be looking to see if the the request asks for JSON+LD, which would be if / when the requester sends an _Accept_ header explicitly asking for JSON+LD. 

The _Accept_ header has an order or precedence, from left-to-right.  A web browser will make some guesses as to the type of content at the URL and create an Accept header like in the screenshot below that guesses various content types and then says `*/*` to mean that "I'll take whatever you give me".  However, other requests, like search engines or consuming apps may be more specific about what they are willing to accept as a content type.

I believe what BBH was getting at is that search engines, or a consuming application, may expect / ask for JSON+LD by sending an _Accept_ header that lists JSON+LD first, or at least before other content types.  That Accept header would look like

`Accept: application/ld+json`

If the request "wants" JSON+LD, then our response should be of _Content-Type: application/ld+json_.

In the Flask Arrest context I believe you'd want to do

`flask_arrest.helpers.serialize_response(json_ld_response_data, 'application/ld+json', ...)
`

Otherwise, if the Accept header doesn't ask for JSON+LD, then we'd return the same info we currently return, which is basic JSON and a response that has

`Content-Type: application/json`

![image](https://user-images.githubusercontent.com/1777776/150623447-edd7d065-3f4a-4151-830d-a2a147786033.png)

**Comment by ramonaspence** [comment]
Okay, I see what I was missing before. 

It looks like Flask-Arrest detects the incoming accept headers and after adding a mimetype with
`@api.outgoing.add_mimetype('application/json+ld')`

I can then define the renderer with:
```
@api.content_renderer.renders('application/json+ld')
     def render_json_ld()
     ... 
 ```
After those two things, I can use `serialize_response()` and specify the renderer, and content-type needed in the serialize_response() method's parameters. 

I believe flask-arrest detects the accept headers itself (with magic) and that confused me in the beginning.

**Comment by allella** [comment]
@BigBlueHat  Ramona's PR for this was merged in.

We've tested with Postman and the JSON also comes up as "valid".

I also ran a validation with https://validator.schema.org but it doesn't seem to query for JSON+LD, so it returns warnings for "Unspecified Type" that the JSON+LD version may resolve.

Thoughts on how to do a proper test of your original search engine friendly feature request?

The current API endpoint for this is https://events.openupstate.org/api/gtc and it will respond to a JSON+LD accepts header.
